File: BOJ/CPP/벽 부수고 이동하기 3(16933).cpp
#include <iostream>
#include <queue>

using namespace std;

int N, M, K;
int map[1001][1001];
int is_visit[1001][1001];
int dx[] = {1, -1, 0, 0}, dy[] = {0, 0, 1, -1};
bool is_daytime = true;

typedef struct HumanInfo {
    int x, y, k_cnt, move_cnt;
} HumanInfo;

int find_min_path() {
    queue<HumanInfo> q;
    
    q.push({0, 0, 0, 1});
    is_visit[0][0] = 0;
    
    while(!q.empty()) {
        int q_size = (int)q.size();
        
        while(q_size--) {
            int cur_x = q.front().x;
            int cur_y = q.front().y;
            int cur_k_cnt = q.front().k_cnt;
            int cur_move_cnt = q.front().move_cnt;
            
            q.pop();
            
            if(cur_x == M - 1 && cur_y == N - 1) {
                return cur_move_cnt;
            }
            
            for(int i = 0; i < 4; i++) {
                int next_x = cur_x + dx[i];
                int next_y = cur_y + dy[i];
                
                if(next_x < 0 || next_y < 0 || next_x >= M || next_y >= N)
                    continue;
                
                if(map[next_y][next_x] == 1) {
                    if(cur_k_cnt >= K)
                        continue;
                    if(is_daytime == false) {
                        q.push({cur_x, cur_y, cur_k_cnt, cur_move_cnt + 1});
                        continue;
                    }
                    if(is_visit[next_y][next_x] > cur_k_cnt + 1) {
                        q.push({next_x, next_y, cur_k_cnt + 1, cur_move_cnt + 1});
                        is_visit[next_y][next_x] = cur_k_cnt + 1;
                        continue;
                    }
                } else {
                    if(is_visit[next_y][next_x] > cur_k_cnt) {
                        q.push({next_x, next_y, cur_k_cnt, cur_move_cnt + 1});
                        is_visit[next_y][next_x] = cur_k_cnt;
                        continue;
                    }
                }
            }
        }
        is_daytime = !is_daytime;
    }
    
    return -1;
}

void set_visit_arr() {
    for(int i = 0; i < N; i++) {
        for(int j = 0; j < M; j++) {
            is_visit[i][j] = 987654321;
        }
    }
}

int main(void) {
    cin >> N >> M >> K;
    
    set_visit_arr();
    
    for(int i = 0; i < N; i++) {
        string str;
        cin >> str;
        for(int j = 0; j < M; j++) {
            map[i][j] = str[j] - '0';
        }
    }
    
    cout << find_min_path() << '\n';
}

File: BOJ/Swift/벽 부수고 이동하기 3(16933).swift
//
//  main.swift
//  Swift_CodingTest
//
//  Created by 피수영 on 2021/07/13.
//

struct Queue<T> {
    private var queue = Array<T>()
    
    public var size: Int {
        return self.queue.count
    }
    
    public var isEmpty: Bool {
        return self.size == 0 ? true : false
    }
    
    mutating public func push(_ value: T) {
        self.queue.append(value)
    }
    
    mutating public func pop() {
        if !self.isEmpty {
            queue.removeFirst()
        }
    }
    
    public func front() -> T? {
        return self.isEmpty == true ? nil : queue.first
    }
}

struct HumanInfo {
    var x: Int
    var y: Int
    var kCount: Int
    var moveCount: Int
    
    init(_ x: Int, _ y: Int, _ kCount: Int, _ moveCount: Int) {
        self.x = x
        self.y = y
        self.kCount = kCount
        self.moveCount = moveCount
    }
}

class Solve {
    var N = Int()
    var M = Int()
    var K = Int()
    var map = [[Int]]()
    var isVisit = [[Int]]()
    let dx: [Int] = [1, -1, 0, 0]
    let dy: [Int] = [0, 0, 1, -1]
    var isDaytime = true

    func findMinPath() -> Int {
        var q = Queue<HumanInfo>()
        
        q.push(HumanInfo(0, 0, 0, 1))
        
        while !q.isEmpty {
            var qSize = q.size
            
            while qSize > 0 {
                qSize -= 1
                
                let curX = q.front()!.x
                let curY = q.front()!.y
                let curKCnt = q.front()!.kCount
                let curMoveCnt = q.front()!.moveCount
                
                q.pop()
                
                if curX == M - 1 && curY == N - 1 {
                    return curMoveCnt
                }
                
                for i in 0..<4 {
                    let nextX = curX + dx[i]
                    let nextY = curY + dy[i]
                    
                    if nextX < 0 || nextY < 0 || nextX >= M || nextY >= N {
                        continue
                    }
                    
                    if map[nextY][nextX] == 1 {
                        if curKCnt >= K {
                            continue
                        }
                        if !isDaytime {
                            q.push(HumanInfo(curX, curY, curKCnt, curMoveCnt + 1))
                            continue
                        }
                        if curKCnt + 1 < isVisit[nextY][nextX] {
                            q.push(HumanInfo(nextX, nextY, curKCnt + 1, curMoveCnt + 1))
                            isVisit[nextY][nextX] = curKCnt + 1
                            continue
                        }
                    } else {
                        if curKCnt < isVisit[nextY][nextX] {
                            q.push(HumanInfo(nextX, nextY, curKCnt, curMoveCnt + 1))
                            isVisit[nextY][nextX] = curKCnt
                            continue
                        }
                    }
                }
            }
            isDaytime.toggle()
        }
        
        return -1
    }

    func main() {
        var numbers = Array(readLine()!.split(separator: " ").map { Int($0)! })
        
        N = numbers[0]
        M = numbers[1]
        K = numbers[2]
        
        map = Array(repeating: Array(repeating: 0, count: M), count: N)
        isVisit = Array(repeating: Array(repeating: 987654321, count: M), count: N)
        
        for i in 0..<N {
            numbers = Array(readLine()!).map { Int(String($0))! }
            for j in 0..<M {
                map[i][j] = numbers[j]
            }
        }
        
        print(findMinPath())
    }
}

let solve = Solve()

solve.main()
